docs/design/services/dynamodb/high-level-library/README.md

```diff
@@ -0,0 +1,38 @@
+# Project Tenets (unless you know better ones)
+
+1. Meeting customers in their problem space allows them to deliver value
+   quickly.
+2. Meeting customer expectations drives usability.
+3. Discoverability drives usage.
+
+# Introduction
+
+This project provides a Dynamo DB experience that is easier to use and
+understand than can be provided by the generated Dynamo DB client. It
+provides object-oriented features, like persisting and updating Java
+data objects, as well as expressive data retrieval operations.
+
+# Project Goals
+
+1. The high-level library should be the ideal way to integrate with
+   Dynamo DB.
+2. The nouns and verbs used by the high-level library should match those
+   in Dynamo DB.
+3. The high-level library should use and build on the idiomatic norms in
+   the rest of the 2.x Java SDK.
+4. The high-level library should perform well in cold JVMs.
+5. Using the low-level Dynamo DB client should be a conscious choice by
+   the customer, not an accidental one.
+   
+# Customer-Requested Changes from 1.11.x
+
+* All built-in Java types should be supported without type converters.
+* There should be just one method of type conversion that can be
+  extended to support any reasonable, foreseeable customer use-case.
+* The high-level library should support immutable object representations
+  from the customer.
+* The high-level library should support object inheritance.
```

Comment:
Does this mean I will be able to do a single `query` operation on a `SuperClass` and get back a `List<SuperClass>` which I can iterate over with polymorphism get two different item types?  For example, my partition key might be order ID and a query on that will get back one item of order metadata and `n` items of order line items.

Reply:
The exact details haven't been shaken out yet, but that's the general idea, yeah. Something like this:

```Java
@DynamoDbSubtype(Dog.class, Cat.class)
class Animal {
    @Attribute 
    @Index(PARTITION_KEY)
    private String id;
}

class Cat extends Animal {
    @Attribute
    private String name;
}

class Dog extends Animal {
    @Attribute
    private String name;
    @Attribute
    private String breed;
}
```

```Java
client.put(new Cat().setId("1").setName("Meowth"));
client.put(new Dog().setId("2").setName("Growlith").setBreed("Awesome"));
client.scan(new Animal()).forEach(animal -> {
    if (animal instanceof Cat) {
        Cat cat = (Cat) animal;
        System.out.println("Cat: " + cat.getName());
    } else if (animal instanceof Dog) {
        Dog dog = (Dog) animal;
        System.out.println("Dog: " + dog.getName() + " (" + dog.getBreed() + ")");
    }
});

Reply:
Awesome!!! ❤️